File: lista5/src/main/scala/Lista5.scala
import Lista5.WeekDay.{Friday, Monday, Saturday, Sunday, Thursday, Tuesday, Wednesday}
//zadanie 1 tylko bazujac na krotkach, typ zminno przeciwnkowy pojedynczej precyzji
//zadanie 1 f odleglosci
//jak zmodyfikowac program aby bylo n-wymiarowa przesten

//zadanie 2 komponowanie struktru, korzystajac z rekordow lub krotek
// scala -> klase
//  typ iloczynowy Person osoba stanowi zlozenie
// koljeny typ to partnership kompozycja - zrobic 2 sposoby uzywac krotki 2 sposob klasy
// funkcja destrukturyzuje struktore i odwoluje sie do parametrow pol osoby

//zadanie 3
// typ sumaryczny reporezuntujacy dzien tyognodnia type dla ocaml
// enum dla scali
// dwie funcje dzialajace na tym typie
// 1 zwroci ciag znakow dla Monday zwrocimy string poniedzialek
// 2 funkcja przyjmuje dzien tygodnia i zwraca koljeny dzien tygodnia
// zadanie 4
// Typ Option wlasna opcja
//zadanie 5 skonstuj solidFigure type
object Lista5 extends App {
  //zadanie 1
  case class Point2D(x: Double, y: Double)

  val first = Point2D(0.0, 0.0)
  val second = Point2D(3.0, 4.0)

  def distance(point1: Point2D, point2: Point2D): Double = math.sqrt(math.pow((point1.x - point2.x), 2) + math.pow(point1.y - point2.y, 2))

  def distance2(point1: (Double, Double), point2: (Double, Double)): Double = math.sqrt(math.pow((point1._1 - point2._1), 2) + math.pow(point1._2 - point2._2, 2))

  case class PointGeneral(arr: List[Double]);
  val pointOne1 = PointGeneral(List(0.0, 0.0, 0.0));
  val pointOne2 = PointGeneral(List(1.0, 1.0, 1.0));

  def distanceGeneral(p1: PointGeneral, p2: PointGeneral): Double = help(p1.arr, p2.arr, 0.0)

  def help(p1: List[Double], p2: List[Double], sum: Double): Double = {
    if (p1.length != p2.length) -9999.0;
    else {
      (p1, p2) match {
        case (Nil, Nil) => math.sqrt(sum)
        case (h1 :: t1, h2 :: t2) => help(t1, t2, sum + math.pow((h2 - h1), 2))
      }
    }
  }

  println(distanceGeneral(pointOne1, pointOne2));
  println(distance2((0.0, 0.0), (3.0, 4.0)))
  println(distance(first, second))


  //zadanie 2
  case class Person(name: String, lastName: String, age: Int, gender: Boolean, shoeSize: Int);
  val per1 = Person("Dawid", "Rym", 20, true, 44);
  val per2 = Person("Ann", "Sym", 22, true, 38);

  case class Partnership(par1: Person, par2: Person);
  val partners = Partnership(per1, per2);

  def whoIsYounger(par: Partnership): String = if (par.par1.age > par.par2.age) par.par2.name else par.par1.name

  println("younger is: " + whoIsYounger(partners));

  val personFunc = (firstName: String) => (lastName: String) => (age: Int) => (gender: Boolean) => (shoeSize: Int) =>
    (firstName, lastName, age, gender, shoeSize);
  val person1 = personFunc("Dawid")("Rym")(20)(true)(44);
  val person2 = personFunc("Daniel")("Hasz")(19)(true)(40);

  val partnership2 = (person1, person2);

  def whoYoungerTuple(tup1: ((String, String, Int, Boolean, Int), (String, String, Int, Boolean, Int))) = {
    val ((firstname1, _, age1, _, _), (firstname2, _, age2, _, _)) = tup1
    if (age1 > age2) firstname2
    else firstname1
  }

  println(whoYoungerTuple(partnership2))


  //zadanie 3
  enum WeekDay {
    case Monday
    case Tuesday
    case Wednesday
    case Thursday
    case Friday
    case Saturday
    case Sunday
  }

  val friday = Friday;

  def weekDatToString(day: WeekDay) = {
    day match {
      case Monday => "Pon"
      case Tuesday => "Wtr"
      case Wednesday => "Srd"
      case Thursday => "Czw"
      case Friday => "Piat"
      case Saturday => "Sob"
      case Sunday => "Niedz"
    }
  }

  def nextDay(day: WeekDay) = {
    day match {
      case Monday => Tuesday
      case Tuesday => Wednesday
      case Wednesday => Thursday
      case Thursday => Friday
      case Friday => Saturday
      case Saturday => Sunday
      case Sunday => Monday
    }
  }

  println(weekDatToString(friday));
  println(nextDay(friday));

  //zadanie 4
    trait Maybe;
    case class Just[A](value: A) extends Maybe;
    case class Nothing() extends Maybe;
//  enum Maybe {
//    case Just[A](value: A)
//    case Nothing
//  }

  def safeHead[A](arr: List[A]): Maybe = {
    arr match {
      case h :: t => Just(h);
      case Nil => Nothing();
    }
  }

  println(safeHead(List(1, 2, 3, 4)))
  println(safeHead(List()))
  //zadanie 5
//  enum SolidFigure {
//    case Cube(x: Double, y: Double, z: Double)
//    case Cone(r: Double, h: Double)
//    case Sphere(r: Double)
//    case Cylinder(r: Double, h: Double)
//  }
  trait Figure;
  case class Cuboid(x: Double, y: Double, z: Double) extends Figure;
  case class Cone(r: Double, h: Double) extends Figure;
  case class Sphere(r: Double) extends Figure;
  case class Cylinder(r: Double, h: Double) extends Figure;
  def volume(figure: Figure) = {
    figure match {
      case Cuboid(x, y, z) => x * y * z;
      case Cylinder(r, h) => Math.PI * math.pow(r, 2) * h
      case Sphere(r) => (4 * Math.PI * math.pow(r, 2)) / 3
      case Cone(r, h) => (Math.PI * math.pow(r, 2) * h) / 3
    }
  }
  val kula = Sphere(1.0)
  println(volume(kula));
  val walec = Cylinder(1.0, 1.0);;
  println(volume(walec));
  val prosto = Cuboid(1.0, 1.0, 1.0);;
  println(volume(prosto));
  val stozek = Cone(1.0, 1.0);;
  println(volume(stozek));
}

File: lista5/lista5.ml
(* zadanie 1 *)
let point2D (x: float) (y: float) = (x, y);;
let point1 = point2D 0. 0.;;
let point2 = point2D 3. 4.;;
let distanceWithTuple a b = let (oSXa, osYa) = a and (osXb, osYb) = b 
in sqrt((osYb -. osYa) ** 2. +. (osXb -. oSXa) ** 2.);;
distanceWithTuple point1 point2;;

type point2DType = {x: float; y: float};;
let pointr1 = {x = 0.; y = 0.};;
let pointr2 = {x = 3.; y = 4.};;
let distanceWithType a b = sqrt((b.y -. a.y) ** 2. +. (b.x -. a.x) ** 2.);;
distanceWithType pointr1 pointr2;;


type pointGeneral = {arr: float list};;
let point31 = {arr = [0.; 0.; 0.]};;
let point32 = {arr = [1.; 1.; 1.]};;
let distanceGeneral a b = if  List.length a.arr != List.length b.arr then None else
  let rec distanceHelper arr1 arr2  sum =
    match (arr1, arr2) with
    | ([], []) -> sqrt(sum)
    | (h1::t1, h2::t2) -> distanceHelper t1 t2 (sum +. (h2 -. h1) ** 2.) in
    distanceHelper a.arr b.arr 0.;;

distanceGeneral point31 point32;;


(* zadanie2 types*)
(* gender true -> male else female *)

type personType = {
  firstname: string; 
  lastname: string; 
  age: int; 
  gender: bool; 
  shoe: int
};;

let personType1 = {
  firstname = "Jan";
  lastname= "Kowalski";
  age = 10;
  gender = true;
  shoe = 43;
};;

let personType2 = {
  firstname = "Piotr";
  lastname = "Nowak";
  age = 11;
  gender = true;
  shoe = 39;
};;
type partnership = {per1: personType; per2: personType};;
let partners = {per1 = personType1; per2 = personType2};;
let whoIsYounger persons = if persons.per1.age < persons.per2.age then persons.per1.firstname else persons.per2.firstname;;
whoIsYounger partners;;


(* zadanie2 tuple  *)
let person (firstname: string) (lastname: string) (age: int) (gender: bool) (shoeSize: int) 
= (firstname, lastname, age, gender, shoeSize);;

let person1 = person "Dawid" "Rymarczyk" 20 true 44;;
let person2 = person "Anja" "Kowalska" 22 false 38;;

let partnership per1 per2 = (per1, per2);;
let given_partners = partnership person1 person2;;

let whoIsYounger people = let ((firstname1,_,age1,_,_), (firstname2,_,age2,_,_)) = people in if age1 < age2 
  then firstname1 else firstname2;;
whoIsYounger given_partners;;


(* zadanie3 *)

type weekDay = Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday;;
let friday = Friday;;

let weekDToString day = 
  match day with
  | Monday -> "Poniedziałek"
  | Tuesday -> "Wtorek"
  | Wednesday -> "Środa"
  | Thursday -> "Czwartek"
  | Friday -> "Piątek"
  | Saturday -> "Sobota"
  | Sunday -> "Niedziela";;
  
weekDToString friday;;


let nextDay day = 
  match day with
  | Monday -> Tuesday
  | Tuesday -> Wednesday
  | Wednesday -> Thursday
  | Thursday -> Friday
  | Friday -> Saturday
  | Saturday -> Sunday
  | Sunday -> Monday;;
nextDay friday;;

(* zadanie4 *)

type 'a maybe = Just of 'a | Nothing;;
let safeHead arr =
  match arr with
  | [] -> Nothing
  | h::t -> Just(h);;

safeHead [];;
safeHead [5;4;3;2;1];;


(* zadanie5 *)
type cuboid = { x: float; y: float; z: float };;
type cone = { r: float; h: float };; 
type sphere = { r: float };;
type cylinder = { r: float; h: float };;
type 'a solidFigure = Cylinder of cylinder | Sphere of sphere | Cone of cone | Cuboid of cuboid;;

let volume figure =
  match figure with
  | Cuboid c -> c.x *. c.y *. c.z
  | Cone c -> (3.14 *. c.r ** 2. *. c.h) /. 3.
  | Sphere b -> (4. *. 3.14 *. b.r ** 2.) /. 3.
  | Cylinder c -> 3.14 *. c.r ** 2. *. c.h;;


let kula = Sphere({r = 1.});;
volume kula;;
let walec = Cylinder({r = 1.; h = 1.});;
volume walec;;
let prosto = Cuboid({x = 1.; y = 1.; z = 1.});;
volume prosto;;
let stozek = Cone({r = 1.; h = 1.});;
volume stozek;;